src/csharp/Grpc.HealthCheck/HealthServiceImpl.cs

```diff
@@ -86,17 +111,124 @@ public void ClearAll()
+                    {
+                        // Remove the response stream from the watchers
+                        if (serverStreamWriters.Remove(responseStream))
+                        {
+                            // Remove empty collection if service has no more response streams
+                            if (serverStreamWriters.Count == 0)
+                            {
+                                watchers.Remove(service);
+                            }
+                        }
+                    }
+                }
+
+                // Allow watch method to exit.
+                watchTcs.TrySetResult(null);
+            });
 
+            // Wait for call to be cancelled before exiting.
+            await watchTcs.Task;
+        }
+
+        private HealthCheckResponse GetHealthCheckResponse(string service, bool throwOnNotFound)
+        {
+            HealthCheckResponse response = null;
+            lock (myLock)
+            {
                 HealthCheckResponse.Types.ServingStatus status;
                 if (!statusMap.TryGetValue(service, out status))
                 {
-                    // TODO(jtattermusch): returning specific status from server handler is not supported yet.
-                    throw new RpcException(new Status(StatusCode.NotFound, ""));
+                    if (throwOnNotFound)
+                    {
+                        // TODO(jtattermusch): returning specific status from server handler is not supported yet.
+                        throw new RpcException(new Status(StatusCode.NotFound, ""));
+                    }
+                    else
+                    {
+                        status = HealthCheckResponse.Types.ServingStatus.ServiceUnknown;
+                    }
+                }
+                response = new HealthCheckResponse { Status = status };
+            }
+
+            return response;
+        }
+
+        private HealthCheckResponse.Types.ServingStatus GetServiceStatus(string service)
+        {
+            if (statusMap.TryGetValue(service, out HealthCheckResponse.Types.ServingStatus s))
+            {
+                return s;
+            }
+            else
+            {
+                return HealthCheckResponse.Types.ServingStatus.ServiceUnknown;
+            }
+        }
+
+        private void NotifyStatus(string service, HealthCheckResponse.Types.ServingStatus status)
+        {
+            if (watchers.TryGetValue(service, out List<IServerStreamWriter<HealthCheckResponse>> serverStreamWriters))
+            {
+                HealthCheckResponse response = new HealthCheckResponse { Status = status };
+
+                foreach (IServerStreamWriter<HealthCheckResponse> serverStreamWriter in serverStreamWriters)
+                {
+                    // TODO(JamesNK): This will fail if a pending write is already in progress.
```

Comment:
the TODO needs to be addressed. I think we should separate the logic of generating the notifications (which should happen under a lock) and have a separate piece of code that tries to send the notifications via the serverStreamWriters (and handles removal or cancelled watch calls).

Reply:
I can add a separate lock that is used when updating and accessing watchers. That doesn't solve this TODO though.

Two issues here:
1. The existing SetStatus/RemoveStatus/ClearAll public methods aren't async.
2. ServerStreamWriter.WriteAsync doesn't allow multiple parallel callers.

Couple of options:
1. Add async versions of SetStatus/RemoveStatus/ClearAll and surround notifying status to the write in an SemaphoreSlim lock.
2. Leave sync methods but wrap writer with a type that manages queuing up pending statuses to send.

It seems like trying to write to a ServerStreamWriter from many threads would be a common problem. Has this been solved before?

Reply:
Went with option 2.

watch is now using Channel for pub/sub. This API doesn't support net45. I have limited it to netstandard targets, and added new net462 target to support newer versions of .NET Framework.

net45 will continue to throw Unimplemented when invoked.